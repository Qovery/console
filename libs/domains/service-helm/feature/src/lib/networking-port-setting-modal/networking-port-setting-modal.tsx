import { type HelmPortRequestPortsInner, PortProtocolEnum } from 'qovery-typescript-axios'
import { type FormEvent } from 'react'
import { Controller, FormProvider, useForm } from 'react-hook-form'
import { InputSelect, InputText, LoaderSpinner, ModalCrud } from '@qovery/shared/ui'
import { useKubernetesServices } from '../hooks/use-kubernetes-services/use-kubernetes-services'

export interface NetworkingPortSettingModalProps {
  helmId: string
  port?: HelmPortRequestPortsInner
  onClose: () => void
  onSubmit: (port: HelmPortRequestPortsInner) => void
}

export function NetworkingPortSettingModal({ helmId, port, onClose, onSubmit }: NetworkingPortSettingModalProps) {
  const isEdit = !!port
  const methods = useForm({
    defaultValues: {
      internal_port: `${port?.internal_port ?? ''}`,
      external_port: `${port?.external_port ?? '443'}`,
      service_name: port?.service_name ?? '',
      namespace: port?.namespace ?? '',
      protocol: port ? port.protocol : PortProtocolEnum.HTTP,
      name: port ? port.name : '',
    },
    mode: 'all',
  })
  const { data: kubernetesServices = [], isLoading } = useKubernetesServices({ helmId })

  const { control, watch, setValue } = methods

  const watchInternalPort = watch('internal_port')
  const watchServiceName = watch('service_name')

  const availableServiceNames = kubernetesServices.map(({ metadata: { name } }) => ({
    label: name,
    value: name ?? '',
  }))

  const availableServicePorts = kubernetesServices
    .find(({ metadata: { name } }) => name === watchServiceName)
    ?.service_spec.ports?.map(({ port, name }) => ({
      label: `${port}${name ? ` (${name})` : ''}`,
      value: `${port}`,
    }))

  const availableProtocols = [PortProtocolEnum.HTTP, PortProtocolEnum.GRPC].map((protocol) => ({
    label: protocol,
    value: protocol,
  }))

  const namePatternRules = {
    pattern: {
      value: /^[a-z][-a-z0-9]{1,37}$/,
      message:
        'Name format is invalid. It should start with a lowercase letter and contain only lowercase letters, numbers, or hyphens (length: 1-37).',
    },
    maxLength: {
      value: 38, // Adjusted to 38 because regex allows up to 37 characters, and the first character is mandatory.
      message: 'Name cannot exceed 37 characters.',
    },
  }

  return (
    <FormProvider {...methods}>
      <ModalCrud
        title={isEdit ? 'Edit port' : 'Set port'}
        onSubmit={methods.handleSubmit((data) =>
          onSubmit({
            ...data,
            internal_port: Number(data['internal_port']), // need to cast to Number
            external_port: Number(data['external_port']),
          })
        )}
        onClose={onClose}
        isEdit={isEdit}
        howItWorks={
          <>
            <p>Specify:</p>
            <ul className="list-disc pl-4">
              <li>the service name as defined within your chart</li>
              <li>the namespace (the one where the service is deployed)</li>
              <li>the port that should be exposed publicly</li>
              <li>the protocol</li>
            </ul>
            <br />
            <p>
              Qovery will take care of configuring the Kubernetes ingress, create a domain and configure the TLS for
              you.
            </p>
          </>
        }
      >
        <div className="flex flex-col gap-5">
          {isLoading ? (
            <div className="flex justify-center">
              <LoaderSpinner />
            </div>
          ) : (
            <>
              {availableServiceNames.length > 0 ? (
                <Controller
                  name="service_name"
                  control={control}
                  rules={{
                    required: 'Please enter service name.',
                  }}
                  render={({ field, fieldState: { error } }) => (
                    <InputSelect
                      label="Service name"
                      value={field.value}
                      options={availableServiceNames}
                      error={error?.message}
                      onChange={(e) => {
                        const name = `p${watchInternalPort}-${e}`
                        setValue('name', name.slice(0, namePatternRules.maxLength.value))
                        setValue('internal_port', '')
                        field.onChange(e)
                      }}
                      portal
                      isSearchable
                    />
                  )}
                />
              ) : (
                <Controller
                  name="service_name"
                  control={control}
                  rules={{
                    required: 'Please enter service name.',
                  }}
                  render={({ field, fieldState: { error } }) => (
                    <InputText
                      name={field.name}
                      onChange={(e: FormEvent<HTMLInputElement>) => {
                        const name = `p${watchInternalPort}-${e.currentTarget.value}`
                        setValue('name', name.slice(0, namePatternRules.maxLength.value))
                        field.onChange(e)
                      }}
                      value={field.value}
                      label="Service name"
                      error={error?.message}
                    />
                  )}
                />
              )}

              {watchServiceName ? (
                availableServicePorts && availableServicePorts.length > 0 ? (
                  <Controller
                    name="internal_port"
                    control={control}
                    rules={{
                      required: 'Please enter an internal port.',
                    }}
                    render={({ field, fieldState: { error } }) => (
                      <InputSelect
                        label="Service port"
                        value={field.value}
                        options={availableServicePorts}
                        error={error?.message}
                        onChange={(e) => {
                          const name = `p${e}-${watchServiceName}`
                          setValue('name', name.slice(0, namePatternRules.maxLength.value))
                          field.onChange(e)
                        }}
                        portal
                        isSearchable
                      />
                    )}
                  />
                ) : (
                  <Controller
                    name="internal_port"
                    control={control}
                    rules={{
                      required: 'Please enter an internal port.',
                      pattern: {
                        value: /^[0-9]+$/,
                        message: 'Please enter a number.',
                      },
                      max: {
                        value: 65535,
                        message: 'Port number must be less than or equal to 65535.',
                      },
                    }}
                    render={({ field, fieldState: { error } }) => (
                      <InputText
                        type="number"
                        name={field.name}
                        onChange={(e: FormEvent<HTMLInputElement>) => {
                          const name = `p${e.currentTarget.value}-${watchServiceName}`
                          setValue('name', name.slice(0, namePatternRules.maxLength.value))
                          field.onChange(e)
                        }}
                        value={field.value}
                        label="Service port"
                        error={error?.message}
                      />
                    )}
                  />
                )
              ) : null}
              <Controller
                name="protocol"
                control={control}
                render={({ field, fieldState: { error } }) => (
                  <InputSelect
                    label="Select protocol"
                    value={field.value}
                    options={availableProtocols}
                    error={error?.message}
                    onChange={field.onChange}
                    portal
                  />
                )}
              />
              <Controller
                name="external_port"
                control={control}
                render={({ field }) => (
                  <InputText
                    type="number"
                    name={field.name}
                    onChange={field.onChange}
                    value="443"
                    label="External port"
                    disabled
                  />
                )}
              />
              <Controller
                name="namespace"
                control={control}
                render={({ field, fieldState: { error } }) => (
                  <InputText
                    name={field.name}
                    onChange={field.onChange}
                    value={field.value}
                    label="Namespace (optional)"
                    error={error?.message}
                  />
                )}
              />
              <div>
                <Controller
                  name="name"
                  defaultValue={watchInternalPort ? `p${watchInternalPort}-${watchServiceName}` : ''}
                  control={control}
                  rules={{
                    required: 'Please enter a port name.',
                    ...namePatternRules,
                  }}
                  render={({ field, fieldState: { error } }) => (
                    <InputText
                      className="mb-1"
                      name={field.name}
                      onChange={field.onChange}
                      value={field.value}
                      label="Port name"
                      error={error?.message}
                    />
                  )}
                />
                <p className="mb-5 ml-3 text-xs text-neutral-350">{`Port Name allows to customize the subdomain assigned to reach the application
port from the internet. Default value is p<port_number>-<service_name>`}</p>
              </div>
            </>
          )}
        </div>
      </ModalCrud>
    </FormProvider>
  )
}

export default NetworkingPortSettingModal
