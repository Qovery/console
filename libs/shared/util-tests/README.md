# shared-util-tests

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test shared-util-tests` to execute the unit tests via [Jest](https://jestjs.io).
