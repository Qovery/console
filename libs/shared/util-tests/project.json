{
  "name": "shared-util-tests",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/shared/util-tests/src",
  "projectType": "library",
  "tags": ["scope:shared", "type:util"],
  "implicitDependencies": ["!shared-ui"],
  "targets": {
    "test": {
      "options": {
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "coverage": true
        }
      }
    }
  }
}
