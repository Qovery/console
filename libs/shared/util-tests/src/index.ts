export * from './lib/shared-util-tests'
