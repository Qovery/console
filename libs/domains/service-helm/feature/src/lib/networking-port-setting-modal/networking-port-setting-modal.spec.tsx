import selectEvent from 'react-select-event'
import { renderWithProviders, screen } from '@qovery/shared/util-tests'
import { NetworkingPortSettingModal } from './networking-port-setting-modal'

jest.mock('../hooks/use-kubernetes-services/use-kubernetes-services', () => ({
  ...jest.requireActual('../hooks/use-kubernetes-services/use-kubernetes-services'),
  useKubernetesServices: () => ({
    data: [],
    isLoading: false,
  }),
}))

describe('NetworkingPortSettingModal', () => {
  it('should match snapshot in create state', () => {
    const { baseElement } = renderWithProviders(<NetworkingPortSettingModal onSubmit={jest.fn()} onClose={jest.fn()} />)
    screen.getByText(/create/i)
    expect(baseElement).toMatchSnapshot()
  })
  it('should match snapshot in edit state', () => {
    const { baseElement } = renderWithProviders(
      <NetworkingPortSettingModal
        port={{
          internal_port: 1234,

          external_port: 443,
          service_name: 'My service',
          namespace: 'My namespace',
          protocol: 'HTTP',
          name: 'My service-p1234',
        }}
        onSubmit={jest.fn()}
        onClose={jest.fn()}
      />
    )
    screen.getByText(/edit/i)
    expect(baseElement).toMatchSnapshot()
  })
  it('should close on cancel', async () => {
    const onClose = jest.fn()
    const { userEvent } = renderWithProviders(<NetworkingPortSettingModal onSubmit={jest.fn()} onClose={onClose} />)
    await userEvent.click(screen.getByText(/cancel/i))
    expect(onClose).toHaveBeenCalled()
  })
  it('should submit port', async () => {
    const onSubmit = jest.fn()
    const { userEvent } = renderWithProviders(<NetworkingPortSettingModal onSubmit={onSubmit} onClose={jest.fn()} />)
    await userEvent.click(screen.getByText(/cancel/i))

    await userEvent.type(screen.getByLabelText(/service name/i), 'foo')
    await userEvent.type(screen.getByLabelText(/namespace/i), 'bar')
    await userEvent.type(screen.getByLabelText(/service port/i), '1234')
    await selectEvent.select(screen.getByLabelText(/protocol/i), ['GRPC'], {
      container: document.body,
    })
    await userEvent.click(screen.getByText(/create/i))

    expect(onSubmit).toHaveBeenCalledWith({
      service_name: 'foo',
      namespace: 'bar',
      internal_port: 1234,
      external_port: 443,
      protocol: 'GRPC',
      name: 'p1234-foo',
    })
  })
})
