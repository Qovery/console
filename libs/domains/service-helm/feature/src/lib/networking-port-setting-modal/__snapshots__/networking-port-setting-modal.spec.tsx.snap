// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkingPortSettingModal should match snapshot in create state 1`] = `
<body>
  <div>
    <div
      class="p-5"
    >
      <h2
        class="h4 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
      >
        Set port
      </h2>
      <span
        aria-controls="radix-:r6:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="mt-2 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
        data-state="closed"
        type="button"
      >
        Show how it works 
        <i
          aria-hidden="true"
          class="fa-regular fa-circle-question text-xs"
        />
      </span>
      <form
        class="mt-6"
      >
        <div
          class="flex flex-col gap-5"
        >
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input     "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Service name"
                  >
                    Service name
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="Service name"
                    name="service_name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
          >
            <div
              class="input input--select    "
              data-testid="select"
            >
              <label
                class="!translate-y-0 !text-xs"
                for="Select protocol"
              >
                Select protocol
              </label>
              <div
                class=" css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-2-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="input-select__control css-13cymwt-control"
                >
                  <div
                    class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
                  >
                    <span
                      class="mr-1 text-sm text-neutral-400"
                    >
                      HTTP
                    </span>
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-readonly="true"
                      class="css-mohuvp-dummyInput-DummyInput"
                      id="Select protocol"
                      inputmode="none"
                      role="combobox"
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <input
                  name="Select protocol"
                  type="hidden"
                  value="HTTP"
                />
              </div>
              <input
                name="Select protocol"
                type="hidden"
                value="HTTP"
              />
              <div
                class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
              >
                <span
                  class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                  role="img"
                />
              </div>
            </div>
          </div>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input input--disabled input--disabled !border-neutral-250  input--label-up "
              >
                <div
                  class="pointer-events-none"
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="External port"
                  >
                    External port
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    disabled=""
                    id="External port"
                    name="external_port"
                    type="number"
                    value="443"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input     "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Namespace (optional)"
                  >
                    Namespace (optional)
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="Namespace (optional)"
                    name="namespace"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="mb-1"
              data-testid="input-text-wrapper"
            >
              <div
                class="relative"
              >
                <div
                  aria-label="input-container"
                  class="input     "
                >
                  <div
                    class=""
                  >
                    <label
                      class="translate-y-2 text-sm"
                      for="Port name"
                    >
                      Port name
                    </label>
                    <input
                      class="input__value "
                      data-testid="input-text"
                      id="Port name"
                      name="name"
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
            <p
              class="mb-5 ml-3 text-xs text-neutral-350"
            >
              Port Name allows to customize the subdomain assigned to reach the application
port from the internet. Default value is p&lt;port_number&gt;-&lt;service_name&gt;
            </p>
          </div>
        </div>
        <div
          class="mt-6 flex justify-end gap-3"
        >
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
            data-testid="cancel-button"
            type="button"
          >
            Cancel
          </button>
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Create
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;

exports[`NetworkingPortSettingModal should match snapshot in edit state 1`] = `
<body>
  <div>
    <div
      class="p-5"
    >
      <h2
        class="h4 max-w-sm truncate text-neutral-400 dark:text-neutral-50"
      >
        Edit port
      </h2>
      <span
        aria-controls="radix-:rd:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="mt-2 cursor-pointer text-sm font-medium text-brand-500 transition hover:text-brand-600"
        data-state="closed"
        type="button"
      >
        Show how it works 
        <i
          aria-hidden="true"
          class="fa-regular fa-circle-question text-xs"
        />
      </span>
      <form
        class="mt-6"
      >
        <div
          class="flex flex-col gap-5"
        >
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input    input--label-up "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Service name"
                  >
                    Service name
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="Service name"
                    name="service_name"
                    type="text"
                    value="My service"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input    input--label-up "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Service port"
                  >
                    Service port
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="Service port"
                    name="internal_port"
                    type="number"
                    value="1234"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
          >
            <div
              class="input input--select    "
              data-testid="select"
            >
              <label
                class="!translate-y-0 !text-xs"
                for="Select protocol"
              >
                Select protocol
              </label>
              <div
                class=" css-b62m3t-container"
              >
                <span
                  class="css-1f43avz-a11yText-A11yText"
                  id="react-select-3-live-region"
                />
                <span
                  aria-atomic="false"
                  aria-live="polite"
                  aria-relevant="additions text"
                  class="css-1f43avz-a11yText-A11yText"
                />
                <div
                  class="input-select__control css-13cymwt-control"
                >
                  <div
                    class="input-select__value-container input-select__value-container--has-value css-1fdsijx-ValueContainer"
                  >
                    <span
                      class="mr-1 text-sm text-neutral-400"
                    >
                      HTTP
                    </span>
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-haspopup="true"
                      aria-readonly="true"
                      class="css-mohuvp-dummyInput-DummyInput"
                      id="Select protocol"
                      inputmode="none"
                      role="combobox"
                      tabindex="0"
                      value=""
                    />
                  </div>
                  <div
                    class="input-select__indicators css-1hb7zxy-IndicatorsContainer"
                  >
                    <span
                      class="input-select__indicator-separator css-1u9des2-indicatorSeparator"
                    />
                    <div
                      aria-hidden="true"
                      class="input-select__indicator input-select__dropdown-indicator css-1xc3v61-indicatorContainer"
                    >
                      <svg
                        aria-hidden="true"
                        class="css-tj5bde-Svg"
                        focusable="false"
                        height="20"
                        viewBox="0 0 20 20"
                        width="20"
                      >
                        <path
                          d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
                <input
                  name="Select protocol"
                  type="hidden"
                  value="HTTP"
                />
              </div>
              <input
                name="Select protocol"
                type="hidden"
                value="HTTP"
              />
              <div
                class="pointer-events-none absolute right-4 top-1/2 -translate-y-1/2"
              >
                <span
                  class="icon-solid-angle-down shrink-0 text-sm text-neutral-400"
                  role="img"
                />
              </div>
            </div>
          </div>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input input--disabled input--disabled !border-neutral-250  input--label-up "
              >
                <div
                  class="pointer-events-none"
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="External port"
                  >
                    External port
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    disabled=""
                    id="External port"
                    name="external_port"
                    type="number"
                    value="443"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class=""
            data-testid="input-text-wrapper"
          >
            <div
              class="relative"
            >
              <div
                aria-label="input-container"
                class="input    input--label-up "
              >
                <div
                  class=""
                >
                  <label
                    class="translate-y-2 text-sm"
                    for="Namespace (optional)"
                  >
                    Namespace (optional)
                  </label>
                  <input
                    class="input__value "
                    data-testid="input-text"
                    id="Namespace (optional)"
                    name="namespace"
                    type="text"
                    value="My namespace"
                  />
                </div>
              </div>
            </div>
          </div>
          <div>
            <div
              class="mb-1"
              data-testid="input-text-wrapper"
            >
              <div
                class="relative"
              >
                <div
                  aria-label="input-container"
                  class="input    input--label-up "
                >
                  <div
                    class=""
                  >
                    <label
                      class="translate-y-2 text-sm"
                      for="Port name"
                    >
                      Port name
                    </label>
                    <input
                      class="input__value "
                      data-testid="input-text"
                      id="Port name"
                      name="name"
                      type="text"
                      value="My service-p1234"
                    />
                  </div>
                </div>
              </div>
            </div>
            <p
              class="mb-5 ml-3 text-xs text-neutral-350"
            >
              Port Name allows to customize the subdomain assigned to reach the application
port from the internet. Default value is p&lt;port_number&gt;-&lt;service_name&gt;
            </p>
          </div>
        </div>
        <div
          class="mt-6 flex justify-end gap-3"
        >
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 hover:[&:not(:focus-visible):not(:disabled)]:shadow-sm active:[&:not(:focus-visible)]:shadow-none outline-brand-500 text-sm h-11 px-5 rounded border border-transparent text-neutral-350 hover:[&:not(:active):not(:focus-visible)]:border-neutral-300 hover:[&:not(:disabled)]:text-neutral-400 dark:hover:[&:not(:disabled)]:text-neutral-250 focus-visible:text-neutral-400 active:bg-neutral-150 dark:active:bg-transparent active:text-neutral-400 disabled:text-neutral-300 disabled:bg-neutral-150 disabled:border-none"
            data-testid="cancel-button"
            type="button"
          >
            Cancel
          </button>
          <button
            class="inline-flex items-center shrink-0 font-medium transition focus-visible:[&:not(:active)]:outline-2 outline-0 shadow-sm disabled:shadow-none active:shadow-none outline-neutral-500 text-sm h-11 px-5 rounded bg-brand-500 active:bg-brand-600 hover:bg-brand-400 text-white disabled:text-brand-300 disabled:bg-brand-100"
            data-testid="submit-button"
            disabled=""
            type="submit"
          >
            Confirm
          </button>
        </div>
      </form>
    </div>
  </div>
</body>
`;
